openid-sharedsignals-framework-1_0.md

```diff
@@ -602,6 +602,19 @@ authorization_schemes
   security considerations, make the authorization_schemes attribute
   publicly accessible without prior authentication.
 
+default_subjects
```

Comment:
@FragLegs @atultulshi Though this will help keep things backward compatible, imo this introduces stream specific behavior being defined in 2 separate places -
1. Config Metadata - Whether all subjects are included in the bootstrapped stream
2. Stream create/update requests - Which events should be part of the stream

Thoughts on [this](https://github.com/openid/sharedsignals/issues/163#issuecomment-2110197678) suggestion?

Reply:
> Do you really think the way a Transmitter treats subjects is stream-specific?
In the case of an IdP (tx) - Service Provider (Rx) relationship, an SP may need to consume events only assigned users. Eg. In case of Duo IdP (Tx) -> Splunk (Rx), Splunk would not want to hear events from non-engineers who never access Splunk beginning from the stream inception.

> It seems strange to modify the stream itself, which should be more constant, in those situations.
My suggestion was only for the stream bootstrap scenario. For future changes, the receiver can continue using add/remove subjects 

> We don't want to return the subjects in the StreamConfiguration data. That would make the PUT and PATCH calls quite difficult if you need to update the stream for any reason.
I agree; we can skip adding subjects to the response. Only keep it in the request to indicate what the stream should bootstrap with. In any case, we do not provide a way for the receiver to know which subjects are part of the stream, Rx is expected to maintain such a state.

> I'm not sure how we would offer the equivalent of the default ALL under your proposal.
All may be shown by the subject as an entire tenant.

Reply:
> In the case of an IdP (tx) - Service Provider (Rx) relationship, an SP may need to consume events only assigned users. Eg. In case of Duo IdP (Tx) -> Splunk (Rx), Splunk would not want to hear events from non-engineers who never access Splunk beginning from the stream inception.

This is a good example for why a Tx might want to default to NONE and let the Rx add specific subjects. What I'm having trouble imagining is that Duo might want to default to NONE for Splunk but then default to ALL for Okta. It seems like an engineering architecture decision (i.e. how you build your transmitter) rather than a stream-by-stream decision.

> My suggestion was only for the stream bootstrap scenario. For future changes, the receiver can continue using add/remove subjects

Having two ways to add subjects seems like it would add a lot of confusion.

> All may be shown by the subject as an entire tenant.

That would still need to be a thing that the Receiver requests in your scenario. It's just that they are requesting it during stream configuration instead of during an add_subject call. What this PR is adding is a way for the Transmitter to say, "By default, we will send all subjects and the Receiver does not have to do anything special." This is the behavior that multiple members of the working group asked for (including you, I thought) during our call.

I'll also note that the interop spec says that "Receivers MUST assume that all subjects are implicitly included in a Stream, without any `AddSubject` method invocations." If Transmitters are going to act that way, I think it would be nice for them to be able to indicate it (although nothing in this PR forces them to indicate it - the new field is entirely optional). I know that other Transmitters do not want to include all subjects implicitly. My interest here is in making that difference visible to a Receiver.

Reply:
I missed this discussion earlier, sorry. @appsdesh , your [suggestion](https://github.com/openid/sharedsignals/issues/163#issuecomment-2110197678) is a good one, but I don't think it is in conflict with this PR. For streams that define their default behavior to be NONE, they could still use the AddSubject in the way you suggest (or add a parameter to the CreateStream method) to specify the locus of subjects they are interested in receiving events about. However I think this is a bigger change (the re-interpretation of AddSubject and the ability to potentially define a subject pattern in CreateStream), so we should take it up after v1.0. Thoughts. Also, I do not think that this PR creates the situation where stream related information is in two places. The Transmitter Configuration Metadata already includes things like "delivery_methods_supported", which are stream related.

Reply:
@FragLegs 

> Duo might want to default to NONE for Splunk but then default to ALL for Okta.

As I understand, SCIM faces a similar issue and solves it by having  _the Transmitter decide_ which subjects are valid on the stream/connection. Our specification already suggests that Rx should reject anything that it does not understand.

@tulshi 

`delivery_methods_supported` is a handshake requirement between a Tx and Rx, we usually won't anticipate the same pair of Tx-Rx doing 2 streams, one on PUSH and other on PULL. So I would not call it a stream property. 

I am ok with not doing _any_ changes related to this and bringing it up post-v1

Reply:
> `delivery_methods_supported` is a handshake requirement between a Tx and Rx, we usually won't anticipate the same pair of Tx-Rx doing 2 streams, one on PUSH and other on PULL. So I would not call it a stream property.

The exact same argument could be made about the proposed `default_subjects` field. It is a handshake requirement in that it defines how the Receiver should expect the Transmitter to perform. It would be _baffling_ for it to be different for 2 streams in the same Tx-Rx pair.

I honestly don't understand your objection to this PR. It is not proposing a new way of managing subjects. It is just trying to make explicit something that already exists but is currently left implicit and undefined.

Perhaps it is confusion about the name of the field. Would it make more sense if the field was called `default_subject_behavior`?

Reply:
> It is a handshake requirement in that it defines how the Receiver should expect the Transmitter to perform. It would be baffling for it to be different for 2 streams in the same Tx-Rx pair.

This is _NOT_ a handshake property, i.e., the absence of this property would not fail the stream establishment, and events would flow. 

A Tx-Rx pair legitimately may have reasons to define one stream with all subjects and another with none, as described above. 

The point I am trying to make is - this is not something that would 
1. Fail the bootstrapping process if it's not available; hence may not be a constituent of discovery/Config Metadata doc.
2. Tx may need to bootstrap streams with various subjects, ALL, NONE, and Subset, based on who the receiver is. Just two values may prevent future flexibility on how we bootstrap streams.